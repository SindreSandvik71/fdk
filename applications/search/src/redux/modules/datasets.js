import _ from 'lodash';
import { addOrReplaceParam } from '../../lib/addOrReplaceUrlParam';
import { fetchActions } from '../fetchActions';

export const DATASETS_REQUEST = 'DATASETS_REQUEST';
export const DATASETS_SUCCESS = 'DATASETS_SUCCESS';
export const DATASETS_FAILURE = 'DATASETS_FAILURE';

function createNestedListOfPublishers(listOfPublishers) {
  const nestedListOfPublishers = _(listOfPublishers).forEach(publisherItem => {
    const filteredChildrenOfParentPublishers = _(listOfPublishers)
      .filter(
        g => g.key.substring(0, g.key.lastIndexOf('/')) === publisherItem.key
      )
      .value();

    filteredChildrenOfParentPublishers.forEach(item => {
      const retVal = item;
      retVal.hasParent = true;
      return retVal;
    });

    const retVal = publisherItem;
    retVal.children = filteredChildrenOfParentPublishers;
    return retVal;
  });

  const resultArray = _(nestedListOfPublishers)
    .filter(f => !f.hasParent)
    .value();

  return resultArray;
}

export function fetchDatasetsIfNeededAction(datasetsURL) {
  // add static size parameter
  const url = addOrReplaceParam(datasetsURL, 'size', '50');
  return (dispatch, getState) => {
    if (!getState().datasets.isFetching) {
      dispatch(
        fetchActions(url, [
          DATASETS_REQUEST,
          DATASETS_SUCCESS,
          DATASETS_FAILURE
        ])
      );
    }
  };
}

const initialState = {
  isFetching: false,
  datasetItems: null,
  publisherCountItems: null
};

export function datasetsReducer(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST: {
      return {
        ...state,
        isFetching: true
      };
    }
    case DATASETS_SUCCESS: {
      const objFromArray = action.payload.aggregations.subjectsCount.buckets.reduce(
        (accumulator, current) => {
          accumulator[current.key] = current; // eslint-disable-line no-param-reassign
          return accumulator;
        },
        {}
      );
      return {
        ...state,
        isFetching: false,
        datasetItems: action.payload,
        publisherCountItems: createNestedListOfPublishers(
          action.payload.aggregations.orgPath.buckets
        ),
        subjectsCountItems: objFromArray
      };
    }
    case DATASETS_FAILURE: {
      return {
        ...state,
        isFetching: false,
        datasetItems: null,
        publisherCountItems: null
      };
    }
    default:
      return state;
  }
}

export const getDatasetById = (datasets, id) => {
  if (datasets && datasets.datasetItems && datasets.datasetItems.hits) {
    return datasets.datasetItems.hits.hits.filter(
      dataset => dataset._source.id === id
    );
  }
  return null;
};
