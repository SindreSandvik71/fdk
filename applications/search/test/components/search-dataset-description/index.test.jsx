import React from 'react';
import { shallow } from 'enzyme';
import DatasetDescription from '../../../src/components/search-dataset-description';
import datasets from '../../fixtures/datasets';

test('should render DatasetDescription correctly with no props', () => {
  const minWrapper = shallow(<DatasetDescription />);
  expect(minWrapper).toMatchSnapshot();
});

test('should render DatasetDescription correctly', () => {
  const { _source } = datasets[0];
  const defaultProps = {
    title: _source.title.nb,
    description: _source.description.nb,
    objective: _source.objective.nb,
    publisher: _source.publisher,
    themes: _source.theme,
    selectedLanguageCode: null,
    harvest: _source.harvest
  };
  const wrapper = shallow(<DatasetDescription {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render DatasetDescription correctly', () => {
  const { _source } = datasets[1];
  const defaultProps = {
    title: _source.title.nb,
    description: _source.description.nb,
    objective: _source.objective.nb,
    publisher: _source.publisher,
    themes: _source.theme,
    selectedLanguageCode: null,
    harvest: _source.harvest
  };
  const wrapper = shallow(<DatasetDescription {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
