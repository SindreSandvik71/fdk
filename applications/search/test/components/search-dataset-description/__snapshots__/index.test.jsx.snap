// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DatasetDescription correctly 1`] = `
<header
  id="dataset-description"
>
  <SideEffect(DocumentMeta)
    description="Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder i Oslo og nærliggende kommuner (markaloven) som trådte i kraft 1. september 2009. Markalovens virkeområde er fastsatt i forskrift 4. september 2015 nr. 1032 om justering av markagrensen fastlegger markalovens geografiske virkeområde med tilhørende kart."
    title="AAMarkagrensen Oslo Kommune og nærliggende kommuner"
  />
  <h1
    className="fdk-margin-bottom"
  >
    AAMarkagrensen Oslo Kommune og nærliggende kommuner
  </h1>
  <div
    className="fdk-detail-date"
  />
  <div
    className="fdk-margin-bottom"
  >
    <span>
      Eies av
       
      <strong
        className="fdk-strong-virksomhet"
        id="dataset-descritption-publisher-text"
      >
        Testetaten
      </strong>
    </span>
    <div
      className="fdk-label fdk-label-on-grey mr-2 mb-2"
      id="dataset-description-theme-undefined"
      key="dataset-description-theme-undefined"
    >
      Forvaltning og offentlig støtte
    </div>
    <div
      className="fdk-label fdk-label-on-grey mr-2 mb-2"
      id="dataset-description-theme-undefined"
      key="dataset-description-theme-undefined"
    >
      Miljø
    </div>
  </div>
  <p
    className="fdk-ingress"
    dangerouslySetInnerHTML={
      Object {
        "__html": null,
      }
    }
  />
  <p
    className="fdk-ingress"
  >
    Datasettes formål er nullam quis rius eget urna mollis ornare vel eu leo. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Vestibulum id ligula porta felis euismod semper.
  </p>
</header>
`;

exports[`should render DatasetDescription correctly 2`] = `
<header
  id="dataset-description"
>
  <SideEffect(DocumentMeta)
    description="Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder i Oslo og nærliggende kommuner (markaloven) som trådte i kraft 1. september 2009. Markalovens virkeområde er fastsatt i forskrift 4. september 2015 nr. 1032 om justering av markagrensen fastlegger markalovens geografiske virkeområde med tilhørende kart."
    title="AAMarkagrensen Oslo Kommune og nærliggende kommuner"
  />
  <h1
    className="fdk-margin-bottom"
  >
    AAMarkagrensen Oslo Kommune og nærliggende kommuner
  </h1>
  <div
    className="fdk-detail-date"
  />
  <div
    className="fdk-margin-bottom"
  >
    <span>
      Eies av
       
      <strong
        className="fdk-strong-virksomhet"
        id="dataset-descritption-publisher-text"
      >
        Testetaten
      </strong>
    </span>
    <div
      className="fdk-label fdk-label-on-grey mr-2 mb-2"
      id="dataset-description-theme-undefined"
      key="dataset-description-theme-undefined"
    >
      Forvaltning og offentlig støtte
    </div>
    <div
      className="fdk-label fdk-label-on-grey mr-2 mb-2"
      id="dataset-description-theme-undefined"
      key="dataset-description-theme-undefined"
    >
      Miljø
    </div>
  </div>
  <p
    className="fdk-ingress"
    dangerouslySetInnerHTML={
      Object {
        "__html": null,
      }
    }
  />
  <p
    className="fdk-ingress"
  >
    Datasettes formål er nullam quis rius eget urna mollis ornare vel eu leo. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Vestibulum id ligula porta felis euismod semper.
  </p>
</header>
`;

exports[`should render DatasetDescription correctly with no props 1`] = `
<header
  id="dataset-description"
>
  <SideEffect(DocumentMeta)
    description=""
    title=""
  />
  <div
    className="fdk-detail-date"
  />
  <div
    className="fdk-margin-bottom"
  />
</header>
`;
